File: src/core/grammar/parse/gss.rb

module GSS

  class GSS
    attr_reader :item, :pos, :edges

    @@nodes = {}
    
    def self.new(*args)
      if !@@nodes.has_key?(args) then
        @@nodes[args] = super(*args)
      else
        @@nodes[args]
      end
      # @@nodes[args] ||= super(*args)
    end

    def self.nodes
      @@nodes
    end
    
    def initialize(item, pos)
      # raise if item.schema_class.name != 'Item'
      @item = item
      @pos = pos
      @edges = {}
      @hash = item.hash * 3 + pos * 17
    end

    def add_edge(node, gss)
      if !edges.include?(node) then
        edges[node] = []
      end
      #edges[node] ||= Set.new
      if edges[node].include?(gss) then
        false
      else
        edges[node] << gss
        true
      end
    end
    
    # def ==(o)
    #   return true if self.equal?(o)
    #   return false unless o.is_a?(GSS)
    #   return item == o.item && pos == o.pos
    # end


    def equals(o)
      if self.equal?(o) then
        true
      elsif !o.is_a?(GSS) then
        false
      else
        item.equals(o.item) && pos == o.pos
      end
    end

    def eql?(o)
      equals(o)
    end

    def hash
      @hash
    end

    def to_s
      "GSS(#{item} @ #{pos})"
    end
  end
end

File: src/js/core/grammar/parse/gss.js
define([
],
function() {
  var GSS ;
  var GSS = MakeClass("GSS", null, [],
    function() {
      this.new = function() {
        var self = this; 
        var args = compute_rest_arguments(arguments, 0);
        if (! self._class_.$.nodes.has_key_P(args)) {
          return self._class_.$.nodes._set(args, super$.new.call(self));
        } else {
          return self._class_.$.nodes._get(args);
        }
      };

      this.nodes = function() {
        var self = this; 
        return self._class_.$.nodes;
      };
    },
    function(super$) {
      this.item = function() { return this.$.item };

      this.pos = function() { return this.$.pos };

      this.edges = function() { return this.$.edges };

      this.initialize = function(item, pos) {
        var self = this; 
        self.$.item = item;
        self.$.pos = pos;
        self.$.edges = new EnsoHash ({ });
        return self.$.hash = item.hash() * 3 + pos * 17;
      };

      this.add_edge = function(node, gss) {
        var self = this; 
        if (! self.edges().include_P(node)) {
          self.edges()._set(node, []);
        }
        if (self.edges()._get(node).include_P(gss)) {
          return false;
        } else {
          self.edges()._get(node).push(gss);
          return true;
        }
      };

      this.equals = function(o) {
        var self = this; 
        if (self.equal_P(o)) {
          return true;
        } else if (! System.test_type(o, GSS)) {
          return false;
        } else {
          return self.item().equals(o.item()) && self.pos() == o.pos();
        }
      };

      this.eql_P = function(o) {
        var self = this; 
        return self.equals(o);
      };

      this.hash = function() {
        var self = this; 
        return self.$.hash;
      };

      this.to_s = function() {
        var self = this; 
        return S("GSS(", self.item(), " @ ", self.pos(), ")");
      };
    });

  GSS = {
    GSS: GSS,

  };
  return GSS;
})
